app/controllers/AccountingPeriodController.scala

```diff
@@ -0,0 +1,119 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import com.google.inject.Inject
+import config.Refined.Max3
+import controllers.actions._
+import forms.DateRangeFormProvider
+import forms.mappings.DateFormErrors
+import models.SchemeId.Srn
+import models.requests.DataRequest
+import models.{DateRange, Mode}
+import navigation.Navigator
+import pages.{AccountingPeriodPage, AccountingPeriods}
+import play.api.data.Form
+import play.api.i18n.{I18nSupport, MessagesApi}
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import services.{SaveService, TaxYearService}
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import uk.gov.hmrc.time.TaxYear
+import utils.FormUtils._
+import utils.RefinedUtils.RefinedIntOps
+import viewmodels.models.DateRangeViewModel
+import views.html.DateRangeView
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AccountingPeriodController @Inject()(
+  override val messagesApi: MessagesApi,
+  navigator: Navigator,
+  identify: IdentifierAction,
+  allowAccess: AllowAccessActionProvider,
+  getData: DataRetrievalAction,
+  requireData: DataRequiredAction,
+  val controllerComponents: MessagesControllerComponents,
+  view: DateRangeView,
+  formProvider: DateRangeFormProvider,
+  saveService: SaveService,
+  taxYearService: TaxYearService
+)(implicit ec: ExecutionContext) extends FrontendBaseController with I18nSupport {
+
+  private def form(usedAccountingPeriods: List[DateRange] = List()) =
+    AccountingPeriodController.form(formProvider, taxYearService.current, usedAccountingPeriods)
+  private val viewModel = AccountingPeriodController.viewModel _
+
+  def onPageLoad(srn: Srn, index: Max3, mode: Mode): Action[AnyContent] = (identify andThen allowAccess(srn) andThen getData andThen requireData) {
+    implicit request =>
+      Ok(view(form().fromUserAnswers(AccountingPeriodPage(srn, index)), viewModel(srn, index, mode)))
+  }
+
+  def onSubmit(srn: Srn, index: Max3, mode: Mode): Action[AnyContent] = (identify andThen allowAccess(srn) andThen getData andThen requireData).async {
+    implicit request =>
+
+      val usedAccountingPeriods = duplicateAccountingPeriods(srn, index)
+
+      form(usedAccountingPeriods).bindFromRequest().fold(
+        formWithErrors => Future.successful(BadRequest(view(formWithErrors, viewModel(srn, index, mode)))),
+        value =>
+          for {
+            updatedAnswers <- Future.fromTry(request.userAnswers.set(AccountingPeriodPage(srn, index), value))
+            _              <- saveService.save(updatedAnswers)
+          } yield Redirect(navigator.nextPage(AccountingPeriodPage(srn, index), mode, updatedAnswers))
+      )
+  }
+
+  def duplicateAccountingPeriods(srn: Srn, index: Max3)(implicit request: DataRequest[_]): List[DateRange] =
+    request.userAnswers.list(AccountingPeriods(srn)).patch(index.arrayIndex, Nil, 1)
```

Comment:
I'm a bit confused here - is this deleting the accounting period at arrayIndex in the saved accounting periods?